include/dsn/internal/message_parser.h

```diff
@@ -87,14 +97,13 @@ namespace dsn
         };
 # endif
 
-        // caller must ensure buffers length is correct as get_send_buffers_count(...);
-        // return buffer count used
-        virtual int prepare_buffers_on_send(message_ex* msg, unsigned int offset, /*out*/ send_buf* buffers) = 0;
-
-        virtual int get_send_buffers_count(message_ex* msg) = 0;
+        // this method will be called before fill_buffers_on_send() to do some prepare operation.
```

Comment:
why two callbacks are needed? why not simply use prepare_on-send as in the old case?